streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java

```diff
@@ -208,9 +209,18 @@ public String toString(final String indent) {
      * @param currentState       the current mapping of {@link HostInfo} -> {@link TopicPartition}s
      * @param clusterMetadata    the current clusterMetadata {@link Cluster}
      */
-    synchronized void onChange(final Map<HostInfo, Set<TopicPartition>> currentState, final Cluster clusterMetadata) {
+    synchronized void onChange(final Map<HostInfo, Set<TaskId>> currentState, final Cluster clusterMetadata, final int version) {
         this.clusterMetadata = clusterMetadata;
-        rebuildMetadata(currentState);
+        rebuildMetadata(currentState, version);
+    }
+
+    private boolean hasPartitionsForAnyTopicPartition(final List<String> topicNames, final Set<String> partitionForHost) {
+        for (final String topicPartition : partitionForHost) {
+            if (topicNames.contains(topicPartition)) {
+                return true;
+            }
+        }
+        return false;
     }
```

Comment:
This function returns true if there is a non-empty intersection between topicNames and partitionForHost. You could reverse the iteration: iterate topicNames (since it is a List) and lookup into partitionForHost (since it is a Set and likely has O(1) lookup time). The overall complexity would be reduced: O(m*n) -> O(m)

Reply:
Makes sense, thanks.